### ISSUE TITLE
Multichapter documents

### ISSUE THREAD

**Issue by Vanuan** [opengovfoundation/madison#653]
Currently, madison doesn't support big documents. There's a hard limit in database. It'd be good to have support of multichapter documents.

**Comment by krues8dr** [comment]
It's true.  Since `doc_contents` is already broken out as a separate table, this would be rather easy to do.  The first step would be to change `contents` from `text` to `mediumtext`.  From there, allowing multiple doc_contents rows per document would make sense, splitting them up on edit and recombining on load.  (Maybe a "view more" on the frontend for each chunk to help users, along with an optional url parameter to get all parts.)

**Comment by cmbirk** [comment]
Agreed.  We probably need to move this issue up as a priority.  When looking at policies that citizens are willing to engage in, these frequently are very long documents or budgets.

**Comment by krues8dr** [comment]
![](http://i.imgur.com/v0TVx1N.gif)

**Comment by Vanuan** [comment]
So, there are actually 2 features I'd like to be supported:
1. To be able to upload large documents
2. To be able to navigate large documents more easily

Solutions:
1. Transparent large document support (store as separate chunks internally, combine at UI side)
2. Arbitrary chapter splitting. Edit and view document chapters independently.

America decoded is great in that regard, but it lacks commenting. E.g.: http://administrative.sanfranciscocode.org/10B/10B.12/

**Comment by krues8dr** [comment]
Your #2 as described is actually pretty tricky, and not something we could do easily, since truly "arbitrary" is, well, arbitrary. That is assuming you'd want something like a tree-view of the hierarchy of the document - the only way to do that is by storing the entire documents as one big XML chunk and doing XPath at the database layer (which is possible in both MySQL and Postgresql these days).  Otherwise we have to reassemble a huge document and xpath it in PHP which is memory-intensive and not great.

It would be far easier to just allow the user to split a document into whatever sections they want and maintain those separately.  That's basically just a new table - `doc_chapter` or similar, which groups the contents together.  We still would have to implement my multiple-content solution above though, because each chapter could also be bigger than `mediumtext` of course.

On State Decoded I've got a few concerns about how we're storing the content, as we're actually splitting out every paragraph into a database row and it's less than ideal.  Though it *does* support commenting via Disqus (see below the content),  it doesn't have annotation.  I've investigated adding annotation to State Decoded via Annotator and it's relatively trivial - but since there are no user accounts in SD, you'd need to use Hypothes.is or build your own.

**Comment by krues8dr** [comment]
A lot of the work has been done here in #832 - but there's still a lot to do.
